From c6f55b3a022f7a08d09496733564ffc127b141bd Mon Sep 17 00:00:00 2001
From: Kishan Kunduru <kkunduru@google.com>
Date: Wed, 15 Jun 2022 23:21:01 +0000
Subject: [PATCH 2/2] curl: Use zerocopy for uplink WAN speedtest

Use zerocopy and increased socket buffer sizes for WAN
Speedtest

BUG: 234028636
TEST: Compiled and verified on sirocco that zerocopy and increased
buffer sizes are used.

Change-Id: I9644b702e851c624661c89f797e2f793adf44746
---
 lib/connect.c | 19 +++++++++++++++++++
 lib/sendf.c   |  6 +++++-
 2 files changed, 24 insertions(+), 1 deletion(-)

diff --git a/lib/connect.c b/lib/connect.c
index 00aca5b45..d56594174 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -1350,6 +1350,25 @@ static CURLcode singleipconnect(struct Curl_easy *data,
                      &rcvlowat, sizeof(rcvlowat)) == -1) {
         failf(data, "Failed to set SO_RCVLOWAT");
       }
+
+      int buffer_len = 8*1024*1024;
+      if (setsockopt(sockfd, SOL_SOCKET, SO_RCVBUF, &buffer_len,
+                     sizeof(buffer_len)) == -1) {
+        failf(data, "Failed to set SO_RCVBUF");
+      }
+    }
+    if(data->set.ul_speedtest_mode) {
+      int one = 0;
+      if (setsockopt(sockfd, SOL_SOCKET, SO_RCVLOWAT,
+                     &one, sizeof(one)) == -1) {
+        failf(data, "Failed to set SO_ZEROCOPY");
+      }
+
+      int buffer_len = 8*1024*1024;
+      if (setsockopt(sockfd, SOL_SOCKET, SO_SNDBUF, &buffer_len,
+                     sizeof(buffer_len)) == -1) {
+        failf(data, "Failed to set SO_SNDBUF");
+      }
     }
   }
 
diff --git a/lib/sendf.c b/lib/sendf.c
index 14ca84bfe..130cc012f 100644
--- a/lib/sendf.c
+++ b/lib/sendf.c
@@ -365,7 +365,11 @@ ssize_t Curl_send_plain(struct Curl_easy *data, int num,
   }
   else
 #endif
-    bytes_written = swrite(sockfd, mem, len);
+    if(data->set.ul_speedtest_mode) {
+      bytes_written = send(sockfd, mem, len, MSG_DONTWAIT|MSG_ZEROCOPY);
+    } else {
+      bytes_written = swrite(sockfd, mem, len);
+    }
 
   *code = CURLE_OK;
   if(-1 == bytes_written) {
-- 
2.37.0.rc0.161.g10f37bed90-goog

